library/alloc/src/vec/mod.rs

```diff
@@ -1001,6 +1001,21 @@ impl<T, A: Allocator> Vec<T, A> {
         }
     }
 
+    /// Converts the vector into [`Box<T>`][owned slice], assuming that
+    /// `self.len() == self.capacity()`.
+    ///
+    /// # Safety
+    /// It's undefined behavior to call this function if `self.capacity() > self.len()`.
+    pub(crate) unsafe fn into_exact_size_boxed_slice(self) -> Box<[T], A> {
```

Comment:
This shouldn't be necessary because `into_boxed_slice` already reuses the allocation when no shrinking is needed.

Reply:
Yup, but it is in fact neccessary since optimizer fails to notice that no shrinking could be needed and doesn’t optimize the checks out.

Reply:
Those are one-time checks, not per iterator item. So it doesn't seem to be worth optimizing them out unless you have some example where it leads to significant code bloat.

Reply:
Previous implementation used just `.into_boxed_slice()` and it led to 2x code bloat (as shown in the original issue):

https://godbolt.org/z/G1jz3q3ba

Reply:
If you turn it into an [u8] it's much shorter, not sure why that is.

Reply:
It works with `u8` (and transparent wrappers over `u8`), but even `u16` leads to the same bloat.

Reply:
Oops, forgot the link.

https://godbolt.org/z/3cxnMvd6E